### PR TITLE
CAT: Validate connector documentation

### LINKED ISSUE #33750: Add CAT scenario testing all required fields are listed under the `Prerequisites` section
## Description

As part of our certification criteria, we are checking that all required fields are listed under the `Prerequisites` section.

## Implementation Details

We need to automatise review the 'Prerequisites' section in the documentation with the fields labeled as 'required' in the spec. This comparison should be meticulous, ensuring that each spec's required fields  is adequately addressed in the prerequisite. We should look for any discrepancies or missing elements, and note any areas where the specification might need adjustments or additional clarifications to align with the prerequisites.

### PR DESCRIPTION
**Pull request by darynaishchenko**
## What
resolved: https://github.com/airbytehq/airbyte/issues/33750 https://github.com/airbytehq/airbyte/issues/33749

## How

Added `TestConnectorDocumentation` with:

- `test_prerequisites_content` - test gets required fields from actual connector spec and check if title is in section under Prerequisites header. Special case is credentials. There are a lot of ways how to describe this step, so for credentials test searches for credentials keywords like auth, account etc.
- `test_docs_structure` - test gets all heading from connector docs and check for the right order based on our standard template. If required heading is not in actual docs or not on a right order test will fail.
- `test_docs_descriptions` - test gets headings that have specific description according our standard template and validates content after this heading to next one line by line. If actual documentation does have differences test will fail.
- `test_validate_links` - retrieves all links from documentation file and makes a get request. if response status code in [200, 403, 401] it's a valid links, other links will be collected as invalid links and showed in assertation.

TestConnectorDocumentation is not required for now.  

## Recommended reading order
1. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py`
2. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/conftest.py`
3. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/config.py`
4. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/doc_templates/` - folder with standard templates to compare with actual docs.

### Manualy tested with:
- `source-google-ads`
- `source-github`
- `source-faker`

### CODE REVIEW

File: airbyte-integrations/bases/connector-acceptance-test/unit_tests/data/docs/correct_all_description_exist.md
Comment (artem1205) [inline airbyte-integrations/bases/connector-acceptance-test/unit_tests/data/docs/correct_all_description_exist.md:38]: Do we need another check for valid URL links? At least assertion for `status_code`

Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/unit_tests/data/docs/correct_all_description_exist.md]: added

**Review by artem1205 — APPROVED** [review]

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1267]: Especially for this test, can you add some comment as to what some of these steps are doing to assemble headers and things might help. It takes a little bit of time to understand what we're pulling from the docs file

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1296]: nit:
```
f" Heading '{heading_names_value}' is not in the right place, the name of heading is incorrect or the heading name is not expected.\n"
```

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1269]: just to confirm, this will get headers at nested levels too right?

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1286]: let's add a comment for found expected header

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1290]: let's add a comment for found a non-required header, moving to next one

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py:14]: why do we stop after Step 2? I presume that many connectors will have steps beyond 2.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1346]: can we add some sort of timeout or failsafe for this? I don't think we want to hold up a CAT run if for example some URLs end up hanging. Maybe 30 seconds (per thread) is a safe amount of time to wait and exit out if it's not fulfilled by then

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1194]: I'm not familiar where the source of this was? Is this officially mentioned by us or just something someone put together? 

I'm just not sure if we want to rely on this structure unless its rooted in something we put out

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1305]: Can we test this some additional certified connectors like `source-greenhouse`, `source-airtable` and a few others. I'm a little concerned how many connectors will fail this validation since it's quite specific to the exact text

**Review by brianjlai — COMMENTED** [review]
I have a feeling that this change might impact some of our existing connectors beyond what you tested, can we make sure to communicate to the team on how to fix these quickly if they fail standard validation?

Can we try to test this with a couple more certified connectors because we should try to be very diligent about how we validate.

Overall nice work so far, but I have a few things I flagged especially when we get very specific to some of the description validations

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1305]: Yes sure, I will test it with some more certified connectors and let you know the results. My previous tests, unfortunately , show differences in the actual and expected docs structure, but it was reasonable, errors like typos, incorrect order of heading etc were found using this test suite.

**Comment by darynaishchenko** [comment]
> I have a feeling that this change might impact some of our existing connectors beyond what you tested, can we make sure to communicate to the team on how to fix these quickly if they fail standard validation?

Yes, this test is really strict and most connectors might fail, but this test suite currently works as TestConnectorAttributes, it doesn't work unless `connector_documentation` is added to acceptance-test-config file.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1269]: no, it gets only top-level headers

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py:14]: thanks, really a good case with steps, as for example netsuite.md has more than 1 steps, we can try to strict compare step 1 and 2 as it is in connector standard template and skip others "Step" headers if they are after "Step 1:" header.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py]: Looking at the basic template and some existing docs, the Step 1, Step 2...etc are all on the second level after `Setup guide`. So if we wanted to test that we have to check nested right?

But I also did a glance over a bunch of certified connector docs and it looks like only a few follow this `Step 1: Set up <source>` and `Step 2: Set up the <source> connector in Airbyte`. Even though this test is opt-in, I think we should just check that `Setup guide` is at the top level as that is the most important part. Whether it follows the sub-header step pattern I don't think is as important

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py
Comment (brianjlai) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/docs.py:15]: see other message about whether we need to test this

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1378]: I don't agree with that, airbyte protocol names it stream so docs should use this naming. Some users can be confused what they need streams or tables.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1371]: docs criteria [here](https://docs.google.com/spreadsheets/d/1Os3--0hRFTrueDoHOi_GuZtpZxApo4iEPZCLentGd98/edit#gid=2122356314), cell E35

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py]: added

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py]: [here](https://docs.google.com/document/d/1mGKXy0fqIZ8sGdyjyU0ps5AVkAapdsUaMuQdyGr6D5E/edit?usp=sharing) is my test results for some connectors

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:1378]: Yes you are right athatconsistent naming scheme is better. No need to address this part.

I did a check and the three sources that use `Supported tables` are `my-hours`, `airtable`, and `dv-360`. I think it is acceptable for these connector docs to fail and to be changed to `Supported streams` to pass tests.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py]: yes, we are validating step 1 and step 2, it just removes Step 1/2 heading name. So `Step 1: Set up <connector name>` str becomes `Set up <connector name>` as in required header. The reason why I removed it that we don't know the step number of this header, `Set up <connector name>`  can be a step 1 or step 2.

**Comment by darynaishchenko** [comment]
Regarding a plan for the team how we will handle failures after this pr will be merged. I will create an issue with a list of cerified connectors and we will add a config for this test in acceptance-config.yaml for these connectors and fix docs in case test are failed. Now this test is not mandatory and will work only if config is provided in acceptance-config.yaml and ql > 400. Connectors without provided config will not be affected.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py
Comment (brianjlai) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py]: I don't have access to this document. Please share with me

**Review by brianjlai — APPROVED** [review]
Please share the test document with me. But I think we can move forward with the test changes

**Review by bazarnov — APPROVED** [review]

**Pull request merged**
